File: adapters/llm/ollama.ts
// adapters/llm/ollama.ts — ILLMClient → Ollama local API (OpenAI-compatible)
import {
  CompletionChunk,
  CompletionRequest,
  CompletionResponse,
  ILLMClient,
  LLMVendor,
} from '../../core/protocols.js'

export interface OllamaConfig {
  model?: string
  maxTokens?: number
  baseUrl?: string
}

interface OllamaChoice {
  message: { role: string; content: string }
  finish_reason: string
}

interface OllamaChatResponse {
  choices: OllamaChoice[]
  usage?: { prompt_tokens: number; completion_tokens: number }
}

export class OllamaLLMClient implements ILLMClient {
  readonly vendor: LLMVendor = 'ollama'
  readonly modelId: string
  readonly maxTokens: number
  readonly supportsTools = false

  private readonly _baseUrl: string

  constructor(config: OllamaConfig = {}) {
    this.modelId = config.model ?? 'llama3.3'
    this.maxTokens = config.maxTokens ?? 4096
    this._baseUrl = config.baseUrl ?? 'http://localhost:11434'
  }

  async complete(request: CompletionRequest): Promise<CompletionResponse> {
    const response = await fetch(`${this._baseUrl}/v1/chat/completions`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(this._buildBody(request)),
    })

    if (!response.ok) {
      throw new Error(`Ollama API error: ${response.status} ${response.statusText}`)
    }

    return this._parse(await response.json() as OllamaChatResponse)
  }

  async *stream(request: CompletionRequest): AsyncIterableIterator<CompletionChunk> {
    const response = await fetch(`${this._baseUrl}/v1/chat/completions`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ ...this._buildBody(request), stream: true }),
    })

    if (!response.ok || !response.body) {
      throw new Error(`Ollama stream error: ${response.status}`)
    }

    const reader = response.body.getReader()
    const decoder = new TextDecoder()

    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      const text = decoder.decode(value)
      for (const line of text.split('\n').filter(l => l.startsWith('data: '))) {
        const json = line.slice(6).trim()
        if (json === '[DONE]') { yield { delta: '', done: true }; return }
        try {
          const event = JSON.parse(json) as {
            choices?: Array<{ delta?: { content?: string } }>
          }
          const delta = event.choices?.[0]?.delta?.content
          if (delta) yield { delta, done: false }
        } catch { /* skip */ }
      }
    }
    yield { delta: '', done: true }
  }

  private _buildBody(request: CompletionRequest): Record<string, unknown> {
    return {
      model: this.modelId,
      messages: request.messages.map(m => ({ role: m.role, content: m.content })),
      ...(request.maxTokens && { max_tokens: request.maxTokens }),
      ...(request.temperature !== undefined && { temperature: request.temperature }),
    }
  }

  private _parse(data: OllamaChatResponse): CompletionResponse {
    const choice = data.choices[0]
    if (!choice) throw new Error('Ollama returned no choices')
    return {
      content: choice.message.content,
      stopReason: choice.finish_reason === 'length' ? 'max_tokens' : 'end_turn',
      usage: {
        inputTokens: data.usage?.prompt_tokens ?? 0,
        outputTokens: data.usage?.completion_tokens ?? 0,
      },
    }
  }
}

File: src/ooagent/adapters/llm/ollama.py
"""adapters/llm/ollama.py — ILLMClient -> Ollama local API (OpenAI-compatible)."""

from __future__ import annotations

import codecs
import json
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any, Literal

import httpx

from ooagent.core.protocols import (
    CompletionChunk,
    CompletionRequest,
    CompletionResponse,
    ILLMClient,
    LLMVendor,
    TokenUsage,
)


@dataclass(frozen=True)
class OllamaConfig:
    model: str | None = None
    max_tokens: int | None = None
    base_url: str | None = None


class OllamaLLMClient(ILLMClient):
    """ILLMClient adapter for the Ollama local API."""

    def __init__(self, config: OllamaConfig | None = None) -> None:
        config = config if config is not None else OllamaConfig()
        self._model_id = config.model if config.model is not None else "llama3.3"
        self._max_tokens = config.max_tokens if config.max_tokens is not None else 4096
        self._base_url = (
            config.base_url if config.base_url is not None else "http://localhost:11434"
        )

    @property
    def vendor(self) -> LLMVendor:
        return "ollama"

    @property
    def model_id(self) -> str:
        return self._model_id

    @property
    def max_tokens(self) -> int:
        return self._max_tokens

    @property
    def supports_tools(self) -> bool:
        return False

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self._base_url}/v1/chat/completions",
                headers={"Content-Type": "application/json"},
                json=self._build_body(request),
            )

        if response.status_code >= 400:
            raise RuntimeError(f"Ollama API error: {response.status_code} {response.reason_phrase}")

        return self._parse(response.json())

    async def stream(self, request: CompletionRequest) -> AsyncIterator[CompletionChunk]:
        body = {**self._build_body(request), "stream": True}
        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST",
                f"{self._base_url}/v1/chat/completions",
                headers={"Content-Type": "application/json"},
                json=body,
            ) as response:
                if response.status_code >= 400:
                    raise RuntimeError(f"Ollama stream error: {response.status_code}")

                decoder = codecs.getincrementaldecoder("utf-8")()
                async for raw in response.aiter_bytes():
                    text = decoder.decode(raw)
                    for line in text.split("\n"):
                        if not line.startswith("data: "):
                            continue
                        payload = line[6:].strip()
                        if payload == "[DONE]":
                            yield CompletionChunk(delta="", done=True)
                            return
                        try:
                            event = json.loads(payload)
                        except json.JSONDecodeError:
                            continue  # skip malformed events
                        choices = event.get("choices") or []
                        delta = (choices[0].get("delta") if choices else None) or {}
                        content = delta.get("content")
                        if content:
                            yield CompletionChunk(delta=content, done=False)
        yield CompletionChunk(delta="", done=True)

    def _build_body(self, request: CompletionRequest) -> dict[str, Any]:
        body: dict[str, Any] = {
            "model": self.model_id,
            "messages": [{"role": m.role, "content": m.content} for m in request.messages],
        }
        if request.max_tokens:
            body["max_tokens"] = request.max_tokens
        if request.temperature is not None:
            body["temperature"] = request.temperature
        return body

    def _parse(self, data: dict[str, Any]) -> CompletionResponse:
        choices = data.get("choices") or []
        if not choices:
            raise RuntimeError("Ollama returned no choices")
        choice = choices[0]
        message = choice.get("message", {})

        finish_reason = choice.get("finish_reason")
        stop_reason: Literal["end_turn", "max_tokens", "tool_use", "stop_sequence"] = (
            "max_tokens" if finish_reason == "length" else "end_turn"
        )

        usage = data.get("usage") or {}
        return CompletionResponse(
            content=message.get("content", ""),
            stop_reason=stop_reason,
            usage=TokenUsage(
                input_tokens=usage.get("prompt_tokens", 0),
                output_tokens=usage.get("completion_tokens", 0),
            ),
        )
